server-spi-private/src/main/java/org/keycloak/policy/HistoryPasswordPolicyProvider.java

```diff
@@ -53,21 +54,23 @@ public PolicyError validate(RealmModel realm, UserModel user, String password) {
         PasswordPolicy policy = session.getContext().getRealm().getPasswordPolicy();
         int passwordHistoryPolicyValue = policy.getPolicyConfig(PasswordPolicy.PASSWORD_HISTORY_ID);
         if (passwordHistoryPolicyValue != -1) {
-            List<CredentialModel> storedPasswords = session.userCredentialManager().getStoredCredentialsByType(realm, user, PasswordCredentialModel.TYPE);
-            for (CredentialModel cred : storedPasswords) {
-                PasswordCredentialModel passwordCredential = PasswordCredentialModel.createFromCredentialModel(cred);
+            List<PasswordCredentialModel> passwordCredentials = session.userCredentialManager()
```

Comment:
Maybe worth to consider this: 
```suggestion
            if (session.userCredentialManager()
                    .getStoredCredentialsByTypeStream(realm, user, PasswordCredentialModel.TYPE)
                    .map(PasswordCredentialModel::createFromCredentialModel)
                    .anyMatch(passwordCredential -> {
                        PasswordHashProvider hash = session.getProvider(PasswordHashProvider.class, passwordCredential.getPasswordCredentialData().getAlgorithm());
                        return hash != null && hash.verify(password, passwordCredential);
                    })) {
                return new PolicyError(ERROR_MESSAGE, passwordHistoryPolicyValue);
            }
```

Reply:
Ahh, yes, this looks a lot better, thanks for the suggestion.